components/support/android/src/main/java/mozilla/appservices/support/native/RustBuffer.kt

```diff
@@ -47,13 +48,18 @@ open class RustBuffer : Structure() {
 
     fun asCodedInputStream(): CodedInputStream? {
         return this.data?.let {
-            CodedInputStream.newInstance(it.getByteBuffer(0, this.len))
+            if (this.len < Int.MIN_VALUE || this.len > Int.MAX_VALUE) {
+                throw RuntimeException("len does not fit in a int")
+            }
+            CodedInputStream.newInstance(it.getByteArray(0, this.len.toInt()))
```

Comment:
Here we avoid using the `ByteBuffer` constructor of `CodedInputStream` since it ends up creating a [buggy implementation](https://github.com/protocolbuffers/protobuf/blob/e667bf6eaaa2fb1ba2987c6538df81f88500d030/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L185-L187).

Reply:
That's unfortunate, as this requires an additional copy. We should fix this when possible...

Reply:
Yup I opened a PR: https://github.com/protocolbuffers/protobuf/pull/7440